### PR TITLE
remove cfssl dependencies

### LINKED ISSUE #80057: CertificateSigningRequest doesn't support URL SANs or isCA
**What would you like to be added**:
I created a CSR with URL-type SANs ( spiffee ), approved it - the generated cert includes the other SANs, but not the URL one.

The bug seems to be in cloudfare.cfssl.ParseCertificateRequest - the template doesn't copy from the CSR.

**Why is this needed**:
It would allow using the K8S CA to generate Istio or other Spiffee certificates, without having to run a separate CA.

Note that as a workaround I tried to create a CSR for an intermediary - but K8S generated a 
cert that doesn't have 'Root=true'. This might be by design - would be good to document what parts of the CSR  are supported, and maybe would be good to allow intermediaries to be signed.
I haven't figured why this is happening.

Including the used CSR and result:
```
-----BEGIN CERTIFICATE REQUEST-----
MIIC9zCCAd8CAQAwLDEqMCgGA1UEAxMhaW50ZXJtZWRpYXJ5LWNhMi5pc3Rpby1z
eXN0ZW0uc3ZjMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA2hArVPoz
QcgORqEXb/tqDhwuDzCDC0U1WNXh3A46c9Sf3o2V6QPEgahxz3/jqyHweBnT9KpC
vSNEltJTPKJEzSfck5BVevtt2TifrKN+zFT98fQZh/OT1jjxoTJtcckunoRFxrb7
RgOAf570SdQMila5Px1TcJK78AQ83+QxuUNOWpsnmlDWglRWrvy83yPlimsZc2mN
el+Vavy5MjuXdvFo+3mD+pxEK94J737mYaxJ5lw3E/+/TZRp2vnAO6uLRti6NXMb
eewy5fl4gHruRaOuedGyI1bQP8hdZ3xwKhjT5PzjOqdRwoU8uFK1IUr/EwayR4Lw
BIXWDRqCLYVUcwIDAQABoIGFMIGCBgkqhkiG9w0BCQ4xdTBzMF0GA1UdEQRWMFSC
EGludGVybWVkaWFyeS1jYTKCHWludGVybWVkaWFyeS1jYTIuaXN0aW8tc3lzdGVt
giFpbnRlcm1lZGlhcnktY2EyLmlzdGlvLXN5c3RlbS5zdmMwEgYDVR0TAQH/BAgw
BgEB/wIBADANBgkqhkiG9w0BAQsFAAOCAQEAo1bdu9/thKgFgdsbgM6dRdIBzXZa
iitted0Fp/s/Jx7jr9Ijemd3sEu/TLpKwOhxoVnlIHoRxtZCuKIEoqqwFWMilPn7
738NcKdN4RIrez43DQfNwXeuxTpQLmCOc64yGXSVLSYfHDlzm0bGgzV60syAl9zA
09enhqJwtZYzGk/Luz3QK0jzv73Q/5oESukiEQjG1+WD/VqU+24LAPlYLbi+q1lc
ATWwR3HtkIMgaKbpS4us+4b+lw50zGEcz+KVnP6G0PdB0M3pSyZJX6kiZxpeHdDC
4+BBUdWrYWV8yxU+2dkaQZNRIL4L2tEmXFg5hsS+fcXnBHPQiBgFjeXXMA==
-----END CERTIFICATE REQUEST-----
openssl req -in csr.pem -noout -text 
... 
X509v3 Basic Constraints: critical
                CA:TRUE, pathlen:0

Generated:
-----BEGIN CERTIFICATE-----
MIIDVTCCAj2gAwIBAgIRAIYi2vf/UfvnpOuSjc66aeMwDQYJKoZIhvcNAQELBQAw
LzEtMCsGA1UEAxMkZTI1YTcwZmQtZjQyYy00YjFjLWI4MTEtNDlmMzU0NGJjNDI5
MB4XDTE5MDcxMTAwMjYwN1oXDTI0MDcwOTAwMjYwN1owLDEqMCgGA1UEAxMhaW50
ZXJtZWRpYXJ5LWNhMi5pc3Rpby1zeXN0ZW0uc3ZjMIIBIjANBgkqhkiG9w0BAQEF
AAOCAQ8AMIIBCgKCAQEA/QaF/Kj6zeJ4qAri9t2mm7dloQMU6XJ39/tzX0cOVI9W
cWS2AzIoTAyTQr7eYsjSUOe9P7pACPOYsXKY8SiLLx8iduiQd4l++0913AgtsGEJ
VjjzDKcO2jZGqNniE1jXFQ0Kc4bGR1xi3HuHy/8P4GQSowo4A9Tn3kcgPynowfnt
/iiLFtDuYXL8eLR2GpMkxbsDQ+Q4unfziCOMumTJUC44N0P76cU5eL7oFENxOfmr
xumOBDGDkhuHmfu38opMDF2UYYP+88B0sg/GeCGFjqNTGvUIk9BmCMNtr2QHm/6V
tSWp88suB5oqksVc7f1rUoqSxF4x1tqOHspE5dTiUwIDAQABo28wbTAMBgNVHRMB
Af8EAjAAMF0GA1UdEQRWMFSCEGludGVybWVkaWFyeS1jYTKCHWludGVybWVkaWFy
eS1jYTIuaXN0aW8tc3lzdGVtgiFpbnRlcm1lZGlhcnktY2EyLmlzdGlvLXN5c3Rl
bS5zdmMwDQYJKoZIhvcNAQELBQADggEBAJS0ARAS7RJBsnFMQ7GSxU850bq+JyX6
1lGeRlFqlhlVBTuXk1CYSh409Xxkab5DrPIIliydwdAEzpld+CxS03tFC+8qFrPv
1PvjYtsYcyQ1IoHfbXRwbVOrVWY6cnTSLLwrv1kjBG7wVTMLhJTLfJdMV1Qqu5nt
84/5m2XTILsyVfg6ew5usp2hUuuPT5WoZ2U1JOD3aBZcK1ecq+OFoFB57t/t2D+7
JM1OiEqsQ/G2eqi66/rC1oHQFIPw2I/bptng0RjfynvRlQwdvo/jMRwcfwUrpdJ8
PLsAiDLkGLvraNrgZCX4iiYNfQN3bbYbDBFGEegnr1x1fIWCOzuQess=
-----END CERTIFICATE-----

```

### PR DESCRIPTION
**Pull request by mikedanese**
remove cfssl dependencies and add support for URI SANs.

/kind bug
/sig auth

Fixes #80057 in part

```release-note
The certificate signer no longer accepts ca.key passwords via the CFSSL_CA_PK_PASSWORD environment variable. This capability was not prompted by user request, never advertised, and recommended against in the security audit.
```

### CODE REVIEW

File: vendor/github.com/cloudflare/cfssl/csr/csr.go [hunk 1]
```diff
@@ -268,6 +261,9 @@ func getHosts(cert *x509.Certificate) []string {
 	for _, email := range cert.EmailAddresses {
 		hosts = append(hosts, email)
 	}
+	for _, uri := range cert.URIs {
```
Comment (mikedanese) [inline vendor/github.com/cloudflare/cfssl/csr/csr.go]: Fix is here.

File: go.mod [hunk 2]
```diff
@@ -346,6 +363,7 @@ replace (
 	github.com/mattn/go-isatty => github.com/mattn/go-isatty v0.0.4
 	github.com/mattn/go-runewidth => github.com/mattn/go-runewidth v0.0.2
 	github.com/mattn/go-shellwords => github.com/mattn/go-shellwords v1.0.5
+	github.com/mattn/go-sqlite3 => github.com/mattn/go-sqlite3 v1.10.0
```
Comment (liggitt) [inline go.mod]: this is a lot of random deps to pick up :-/

I know we're not actually depending on all these packages, but their preferred versions of transitive deps still affect version resolution :-/

File: Godeps/LICENSES [hunk 3]
```diff
@@ -19287,6 +19315,481 @@ SOFTWARE.
...
+BoringSSL test runner. Files that were created by Google, and new files in
+forks of packages maintained by Google have a Google copyright and fall under
+the ISC license. In addition ZCrypto includes a `util/isURL.go` file created by
+Alex Saskevich and licensed under the MIT license. All other files are copyright
+Regents of the University of Michigan, and fall under the Apache 2.0 license.
+All three licenses are reproduced at the bottom of this file.
+
+--------
+
+ISC License used for Google code
```
Comment (liggitt) [inline Godeps/LICENSES]: I don't recognize this license, is it on the approved list?

File: vendor/github.com/weppos/publicsuffix-go/publicsuffix/rules.go [hunk 4]
```diff
@@ -0,0 +1,8786 @@
+// This file is automatically generated
+// Run "go run cmd/gen/gen.go" to update the list.
+
+package publicsuffix
+
+const defaultListVersion = "PSL version 6f03f4 (Fri May 31 16:38:49 2019)"
```
Comment (liggitt) [inline vendor/github.com/weppos/publicsuffix-go/publicsuffix/rules.go]: How often is this updated and what are the consequences of being out of date? Do we fail open or closed on unknown URL suffixes?

File: vendor/github.com/weppos/publicsuffix-go/publicsuffix/rules.go [hunk 5]
```diff
@@ -0,0 +1,8786 @@
...
+
+func DefaultRules() [8767]Rule {
+	return r
+}
+
+var r = [8767]Rule{
```
Comment (liggitt) [inline vendor/github.com/weppos/publicsuffix-go/publicsuffix/rules.go]: does this end up allocating on init of all binaries or just some?

File: vendor/github.com/zmap/zcrypto/x509/README.md [hunk 6]
```diff
@@ -0,0 +1,8 @@
+Originally based on the go/crypto/x509 standard library,
+this package has now diverged enough that it is no longer
+updated with direct correspondence to new go releases.
```
Comment (liggitt) [inline vendor/github.com/zmap/zcrypto/x509/README.md]: This is vaguely concerning. What's the level of comfort with a non-stdlib crypto library?

File: vendor/github.com/zmap/zlint/util/gtld_map.go [hunk 7]
```diff
@@ -0,0 +1,7830 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package util
+
+var tldMap = map[string]GTLDPeriod{
+	"aaa": {
```
Comment (liggitt) [inline vendor/github.com/zmap/zlint/util/gtld_map.go]: same questions about allocation on init of all binaries, frequency of updates, and implications of being outdated

**Comment by mikedanese** [comment]
@liggitt, I'm going to work on removing the cfssl dependency instead. It looks straightforward.

File: pkg/controller/certificates/signer/cfssl_signer.go [hunk 8]
```diff
@@ -1,153 +0,0 @@
...
-	if err != nil {
-		return nil, fmt.Errorf("error reading CA key file %q: %v", caKeyFile, err)
-	}
-
-	parsedCa, err := helpers.ParseCertificatePEM(ca)
-	if err != nil {
-		return nil, fmt.Errorf("error parsing CA cert file %q: %v", caFile, err)
-	}
-
-	strPassword := os.Getenv("CFSSL_CA_PK_PASSWORD")
```
Comment (liggitt) [inline pkg/controller/certificates/signer/cfssl_signer.go]: This is an external-facing change. Do we need to preserve this?

**Comment by liggitt** [comment]
Verify error and one question about the private key password envvar behavior, lgtm otherwise

File: pkg/controller/certificates/signer/cfssl_signer.go (hunk 8 above)
Comment (mikedanese) [reply pkg/controller/certificates/signer/cfssl_signer.go]: I would be ok with a release note if you think anyone is actually using this. This is undocumented behavior, and it seems like it would be unlikely. Happy to preserve it.

**Comment by BenTheElder** [comment]
Pretty sure CI has binaries in an image in test-infra which CI will use
instead

File: pkg/controller/certificates/authority/policies.go [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+// x509.ExtKeyUsage types.
+func keyUsagesFromStrings(usages []capi.KeyUsage) (x509.KeyUsage, []x509.ExtKeyUsage, error) {
+	var keyUsage x509.KeyUsage
+	var extKeyUsage []x509.ExtKeyUsage
+	var unrecognized []capi.KeyUsage
+	for _, usage := range usages {
+		if val, ok := keyUsageDict[usage]; ok {
+			keyUsage |= val
+		} else if val, ok := extKeyUsageDict[usage]; ok {
+			extKeyUsage = append(extKeyUsage, val)
```
Comment (liggitt) [inline pkg/controller/certificates/authority/policies.go]: does this need to dedupe/sort to make output deterministic?

File: pkg/controller/certificates/signer/signer_test.go [hunk 10]
```diff
@@ -0,0 +1,96 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
Comment (liggitt) [inline pkg/controller/certificates/signer/signer_test.go]: nit: 2019 on new files

File: go.mod [hunk 11]
```diff
@@ -15,7 +15,7 @@ require (
 	github.com/GoogleCloudPlatform/k8s-cloud-provider v0.0.0-20190822182118-27a4ced34534
 	github.com/JeffAshton/win_pdh v0.0.0-20161109143554-76bb4ee9f0ab
 	github.com/Microsoft/go-winio v0.4.11
-	github.com/Microsoft/hcsshim v0.0.0-20190417211021-672e52e9209d
```
Comment (liggitt) [inline go.mod]: is this related?

File: pkg/controller/certificates/authority/policies.go [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+// CertificateAuthority. It may default or otherwise mutate a certificate
+// template.
+type SigningPolicy interface {
+	// not-exporting apply forces signing policy implementations to be internal
+	// to this package.
+	apply(template *x509.Certificate) error
+}
+
+// PermissiveSigningPolicy is the signing policy historically used by the local
+// signer. It permits CertificateRequests with any SANs and applies a TTL.
```
Comment (liggitt) [inline pkg/controller/certificates/authority/policies.go]: should we enforce things like "IsCA: false" and no extensions here, rather than in the signer setup of the CR template?

File: pkg/controller/certificates/authority/authority.go [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+		NotBefore:          nbf,
+	}
+	if err := policy.apply(tmpl); err != nil {
+		return nil, err
+	}
+
+	if ca.Certificate.NotAfter.Before(now) {
+		return nil, fmt.Errorf("refusing to sign a certificate that expired in the past")
+	}
+	if ca.Certificate.NotAfter.Before(tmpl.NotAfter) {
```
Comment (liggitt) [inline pkg/controller/certificates/authority/authority.go]: also reject CSRs where NotBefore == NotAfter?

File: pkg/controller/certificates/signer/cfssl_signer.go (hunk 8 above)
Comment (liggitt) [reply pkg/controller/certificates/signer/cfssl_signer.go]: added in https://github.com/kubernetes/kubernetes/commit/fb099ae3856750ddad24a8cc5742b515d033c75d#diff-f0a1084e82f4fa0dd67b7d41dd1d3fd6R63, doesn't look like it was prompted by a specific use case.

also called out as problematic in https://github.com/kubernetes/kubernetes/issues/81117

I'm fine with a release note and removal

File: pkg/controller/certificates/authority/authority.go [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+		return nil, fmt.Errorf("unable to generate a serial number for %s: %v", cr.Subject.CommonName, err)
+	}
+
+	tmpl := &x509.Certificate{
+		SerialNumber:       serialNumber,
+		Subject:            cr.Subject,
+		DNSNames:           cr.DNSNames,
+		IPAddresses:        cr.IPAddresses,
+		EmailAddresses:     cr.EmailAddresses,
+		URIs:               cr.URIs,
```
Comment (liggitt) [inline pkg/controller/certificates/authority/authority.go]: if we're adding support for URIs, also update describer:

https://github.com/kubernetes/kubernetes/blob/7b6369c803cd06f0084ac57b2b6266000b739aea/staging/src/k8s.io/kubectl/pkg/describe/versioned/describe.go#L3264-L3273

and kubelet client cert recognizers to require URI SANs to be empty:

https://github.com/kubernetes/kubernetes/blob/7b6369c803cd06f0084ac57b2b6266000b739aea/pkg/controller/certificates/approver/sarapprove.go#L176-L178

**Comment by liggitt** [comment]
I really like dropping this dep

File: pkg/controller/certificates/authority/policies.go (hunk 12 above)
Comment (mikedanese) [reply pkg/controller/certificates/authority/policies.go]: Ya, I was thinking about it. We might want to reserve some rights in the CA, but I don't see a need for that now.

File: go.mod (hunk 11 above)
Comment (mikedanese) [reply go.mod]: Fixed. I don't know how to use go mod apparently.

File: pkg/controller/certificates/signer/cfssl_signer.go (hunk 8 above)
Comment (mikedanese) [reply pkg/controller/certificates/signer/cfssl_signer.go]: Done.

File: pkg/controller/certificates/authority/policies.go (hunk 12 above)
Comment (liggitt) [reply pkg/controller/certificates/authority/policies.go]: document the extension and ca behavior of this signing policy

**Comment by liggitt** [comment]
one nit on godoc, then lgtm

File: pkg/controller/certificates/authority/policies.go (hunk 12 above)
Comment (mikedanese) [reply pkg/controller/certificates/authority/policies.go]: Done.
